feat: Add support for in cluster generation

bradbeam opened the pull request:
This also includes #382 ( we can either keep included with this PR and close the other or keep separate )

### Reminders
- ~Ensure we cut a new release of Konjure~
- ~Ensure we cut a new release of Optimize-go~

bradbeam committed:
feat: Begin support for in cluster experiment generation

Some basic stubs for discussion, still much to be done.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
feat: Add NewApplicationAPI

This adds in the ability to create a new application api client

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
feat: Build out scan functions

This should add in support for consuming the api representations and combining them
into the controller representation for an application.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
tests: Add basic coverage for scan functions

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Clean up runner code a little

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Update tags to use constants

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
refactor: Split up common low level functions for experiment conversion

This introduces a few helper functions to support some of the common activies in parsing
an in cluster experiment into the API representation.
This should allow us to better support the commonalities between application template an
experiment.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
feat: Add function to convert cluster.experiment to api.template

This introduces a helper function to convert an in cluster experiment into
an application.template.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
feat: Fill in the rest of the activity workflow

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
fix: Minor fixups for application generation

- Create objectives properly ( need to use objective.goals instead of just objectives )
- Set application defaults to have :sparkles: happen properly ( takes care of associating objective name to type )
- Use correct url for updating template ( was incorrectly using scenario url instead of template url )

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
tests: Fix up runner tests

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Remove unused code

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
feat: Add support for updating an in cluster experiment from an api template

This introduces support for updating an in cluster experiment with a previous scan/templates
results.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
feat: Add support for getting baselines

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
core: Fix compilation errors

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
feat: Add support for updating application activity status on failure

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Fix some tests and clean up funcs

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

jgustie commented on internal/experiment/runner.go:
```diff
@@ -0,0 +1,291 @@
+		case <-ctx.Done():
+			return
+		case activity := <-activityCh:
+			// TODO might want to consider moving this to a func so we can defer delete activity
```
I like this idea.

jgustie commented on internal/experiment/runner.go:
```diff
@@ -0,0 +1,291 @@
+}
+
+func (r *Runner) handleErrors(ctx context.Context, err error, activityURL string) {
+	failure := applications.ActivityFailure{FailureMessage: err.Error()}
```
This should have a reason code: generally the error checking looks for a non-empty reason code but doesn't look at the message.

jgustie commented on internal/experiment/runner.go:
```diff
@@ -0,0 +1,291 @@
+	}
+
+	query := applications.ActivityFeedQuery{}
+	query.SetType("poll", applications.TagScan, applications.TagRun)
```
`"poll"` shouldn't be here, it isn't a tag (it's a made up hub type).

jgustie commented on internal/experiment/runner.go:
```diff
@@ -0,0 +1,291 @@
+	}, nil
+}
+
+// This doesnt necessarily need to live here, but seemed to make sense
```
This makes golint sad.

bradbeam replied:
didn't know if runner made sense to keep under `internal/experiment` or move to a separate package. I think initially I had it here so I could manipulate some of the generate functionality, but I don't think that's necessary anymore.

jgustie commented on main.go:
```diff
@@ -135,6 +137,15 @@ func main() {
+	}
+
+	ctx := context.Background()
+	go runner.Run(ctx)
```
This means the runner never gets to shutdown cleanly, I wonder if we should be leveraging `mgr.Add(...)` to get the manager's shutdown message:
```
// Alternately, there is a `manager.RunnableFunc`...
func (r *Runner) Start(stop <-chan struct{}) error {
	ctx, cancel := context.WithCancel(context.Background())
	go func() {
		<-stop
		cancel()
	}()
	r.Run(ctx)
	return nil
}

// ...then you can do
mgr.Add(runner)
```

jgustie commented on controllers/server_controller.go:
```diff
@@ -73,6 +74,7 @@ type ServerReconciler struct {
 	Log            logr.Logger
 	Scheme         *runtime.Scheme
 	ExperimentsAPI experimentsv1alpha1.API
+	// ApplicationsAPI applicationsv1alpha1.API
```
?

bradbeam replied:
left over from trying to do too much at once :oops:

jgustie commented on internal/experiment/scan.go:
```diff
@@ -0,0 +1,201 @@
+	"fmt"
+
+	"github.com/thestormforge/konjure/pkg/konjure"
+	redskyappsv1alpha1 "github.com/thestormforge/optimize-controller/v2/api/apps/v1alpha1"
```
This should be `optimizeappsv1alpha1`.

jgustie commented on internal/experiment/scan.go:
```diff
@@ -0,0 +1,201 @@
+	// Construct a controller representation of an application from the api definition
+	baseApp := &redskyappsv1alpha1.Application{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: scenario.Name,
```
Shouldn't this use `app.Name`?

jgustie commented on internal/experiment/scan.go:
```diff
@@ -0,0 +1,201 @@
+	}
+
+	return kResources, nil
+}
```
I think this can be simplified to just:
```
func (r *Runner) scanResources(app applications.Application, baseApp optimizeappsv1alpha1.Application) error {
  data, err := json.Marshal(app.Resources)
  if err != nil {
    return err
  }
  return json.Unmarshal(data, &baseApp.Resources)
}
```
We do not need to enforce the resource type (even if the application API won't ever give us anything else), and the validation constraints here are more limiting then what is allowed by Konjure (even if the application API can't represent it).

bradbeam replied:
👍 
there's a chance this was left over from before I had updated konjure and had to do a little manual intervention

jgustie commented on internal/experiment/scan.go:
```diff
@@ -0,0 +1,201 @@
+	// fake kubectl output
+	if r.kubectlExecFn != nil {
+		g.FilterOptions = scan.FilterOptions{KubectlExecutor: r.kubectlExecFn}
+	}
```
This can just be `g.FilterOptions.KubectlExecutor = r.kubectlExecFn` (embrace the default values!)

jgustie commented on internal/experiment/scan.go:
```diff
@@ -0,0 +1,201 @@
+
+	appScenario := []redskyappsv1alpha1.Scenario{
+		{
+			// TODO Should this be name or title?
```
Name is correct, we don't have human readable names in `optimizeappsv1alpha1`. But to that point, you should be able to round trip a whole scenario:
```
func (r *Runner) scanScenarios(scn applications.Scenario, baseApp optimizeappsv1alpha1.Application) error {
  data, err := json.Marshal(scn)
  if err != nil {
    return err
  }
  baseApp.Scenarios = make([]optimizeappsv1alpha1.Scenario, 1)
  return json.Unmarshal(data, &baseApp.Scenarios[0])
```
Though this might depend on OPT-767 to make sure all the names are in alignment re: branding.

jgustie commented on internal/server/api.go:
```diff
@@ -19,64 +19,79 @@ package server
+		return nil, err
 	}
 
 	address := strings.TrimSuffix(srv.API.ExperimentsEndpoint, "/v1/experiments/")
```
This is a dumb problem to have, but the configuration model and the API client model diverged when the API version became part of the service path. The unfortunate side effect is that:
```
address := strings.TrimSuffix(srv.API.ExperimentsEndpoint, "/v1/experiments/")
```
Needs to be:
```
address := strings.TrimSuffix(srv.API.ApplicationsEndpoint, "/v2/applications/")
```
Even though they _should_ effectively produce the same value.

bradbeam committed:
chore: Cleanup from review comments

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
refactor: Move experiment/scan to server

This better oragnizes the handling of the api representation of an
application to an in cluster representation of an application.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
refactor: Migrate runner to controllers package

This moves/renames/refactors the runner to be a manager.Runnable so we
can integrate tighter with the controller manager. This should allow us
to better handle shutdowns and better aligns with existing codebase.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
refactor: Use runtime.objects instead of explicit resource types

This switches out explicit resource types ( secret/deployment/experiment/etc ) for a more
generic runtime.Object approach. This simplifies a number of the calls from generating
the experiment in cluster.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Add additional logging statements

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
feat: Move rbac generation from kubebuilder to cli

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Update types to handle field renames

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
fix: Dont treat application api errors as fatal

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

jgustie commented on cli/internal/commands/authorize_cluster/generator.go:
```diff
@@ -295,7 +295,8 @@ func performanceTokens(label, controllerName string, data map[string][]byte) con
 	// List all the organizations the user belongs to
-	orgs, err := exec.Command("forge", "organization", "list", "--output", "plain").Output()
+	// Janky hack mate
+	orgs, err := exec.Command("false", "forge", "organization", "list", "--output", "plain").Output()
```
Should we make StormForger credentials opt-in for now?

bradbeam replied:
This was a workaround(?) because I've got a service account token in my stormforge.toml that I wanted to use. I'm not sure if the service account is limited to not being able to create additional service accounts, but without triggering the err here, I was unable to get the storm forger jwt added to my controller deployment

bradbeam replied:
( it shouldn't have been included here, was a sloppy git add )

jgustie commented on cli/internal/commands/grant_permissions/generator.go:
```diff
@@ -171,6 +175,66 @@ func (o *GeneratorOptions) generateClusterRole(roleRef *rbacv1.RoleRef) *rbacv1.
+				Verbs:     []string{""},
+				APIGroups: []string{""},
+				Resources: []string{""},
+			},
```
Was this a WIP template?

bradbeam replied:
☠️ 
yeah just a copy/paste template

jgustie commented on cli/internal/commands/grant_permissions/generator.go:
```diff
@@ -77,6 +80,7 @@ func (o *GeneratorOptions) addFlags(cmd *cobra.Command) {
 	cmd.Flags().BoolVar(&o.CreateTrialNamespaces, "create-trial-namespace", o.CreateTrialNamespaces, "include trial namespace creation permissions")
 	cmd.Flags().StringVar(&o.NamespaceSelector, "ns-selector", o.NamespaceSelector, "bind to matching namespaces")
 	cmd.Flags().BoolVar(&o.IncludeManagerRole, "include-manager", o.IncludeManagerRole, "bind manager to matching namespaces")
+	cmd.Flags().BoolVar(&o.CreatePrometheusRBAC, "prometheus", true, "roles to create Prometheus in cluster")
```
This should default to the current value in the struct, then to adjust the default when adding the subcommand ([here](https://github.com/thestormforge/optimize-controller/blob/4015d2b82c4acfa4c62a1e7e46df3239cbfc4dd8/cli/internal/commands/generate/generate.go#L44) and [here](https://github.com/thestormforge/optimize-controller/blob/4015d2b82c4acfa4c62a1e7e46df3239cbfc4dd8/cli/internal/commands/initialize/generator.go#L179])).

jgustie commented on internal/server/api.go:
```diff
@@ -18,64 +18,79 @@ package server
 	}
 
-	address := strings.TrimSuffix(srv.API.ExperimentsEndpoint, "/v1/experiments/")
+	address := strings.TrimSuffix(srv.API.ApplicationsEndpoint, "/v2/applications/")
```
This still doesn't address my original point: the way we are setup right now you should not use the same client for the application and experiment APIs. The least impactful hack around this that I can think of right now would be something like this:
```
func newClientFromConfig(ctx context.Context, uaComment string, address func(config.Server) string) (api.Client, error) {
  ...
  c, err := api.NewClient(address(srv), rt)
  ...
}

# From NewExperimentAPI call:
client, err := newClientFromConfig(ctx, uaComment, func(srv configServer) string {
  return strings.TrimSuffix(srv.API.ExperimentsEndpoint, "/v1/experiments/")
})

# From NewApplicationAPI call:
client, err := newClientFromConfig(ctx, uaComment, func(srv configServer) string {
  return strings.TrimSuffix(srv.API. ApplicationsEndpoint, "/v2/applications/")
})
```

bradbeam replied:
they're still different http clients, just both configured to the same endpoint.
but do we want to have them split? I thought since we'll be proxying requests through applications endpoint shouldn't they be using the same endpoint?

jgustie replied:
The issue is that the client was refactored to be shared between the APIs (i.e. the "v1" was removed from the path), but the configuration was not: there are still distinct entry points for experiments and applications represented in the configuration. Technically the two trim calls above _should_ produce the same results if you are taking the defaults, but the config does expose the contract for changing them independently.

jgustie commented on controllers/application_poller_controller.go:
```diff
@@ -0,0 +1,278 @@
+
+func (p *Poller) Start(ch <-chan struct{}) error {
+	if p.apiClient == nil {
+		<-ch
```
Just return immediately and let the Go routine complete?

bradbeam replied:
yeah that'd probably be fine. assuming I misinterpreted what https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.5.0/pkg/manager#RunnableFunc was meaning by `it being important that it blocks until it<I read this as manager, but probably means function> is done`

jgustie replied:
Yeah, the runnables just get sent off into a [Go routine](https://github.com/kubernetes-sigs/controller-runtime/blob/0fcf28efebc9a977c954f00d40af966d6a4aeae3/pkg/manager/internal.go#L225).

jgustie commented on controllers/application_poller_controller.go:
```diff
@@ -0,0 +1,278 @@
+	// We always want to delete the activity after having received it
+	defer func() {
+		if err := p.apiClient.DeleteActivity(ctx, activity.URL); err != nil {
+			p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to delete activity", err), activity.URL)
```
This should probably just log the error instead of trying to update the activity it just failed to delete.

jgustie commented on internal/server/application.go:
```diff
@@ -0,0 +1,300 @@
+		return nil, err
+	}
+
+	// Handle internal apis referencing `stormforger` instead of `stormforge-perf`
```
We should get the app API updated so it's returning the right thing.

jgustie commented on controllers/application_poller_controller.go:
```diff
@@ -0,0 +1,278 @@
+}
+
+func (p *Poller) handleErrors(ctx context.Context, err error, activityURL string) {
+	failure := applications.ActivityFailure{FailureMessage: err.Error()}
```
Still needs a reason or the failure never happened.

bradbeam replied:
Reasons ( scan failed / run failed )

jgustie commented on controllers/application_poller_test.go:
```diff
@@ -0,0 +1,205 @@
+
+	scheme := runtime.NewScheme()
+	redskyv1beta2.AddToScheme(scheme)
+	redskyappsv1alpha1.AddToScheme(scheme)
```
`s/redsky/optimize/g`

bradbeam committed:
chore: Set defaults on in-cluster rbac consistently

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Cleanup from code review

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Update konjure

Think this was accidentally removed in a previous rebase

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
fix: Add missing services permission for in cluster generation

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Add consistent log message when api is unavailable

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Temporary fix for optimize-go

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
fix: dont use address of loop var :oops:

This fixes an issue with using the address of a loop var for
baseline assignments. This would incorrectly set all baselines
to the value of the last loop item and result in a downstream
impact of the baseline value being out of range for the experiment.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
chore: Add comments, fix nil test

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
fix: Use proper type from numorstr/intstr based on parameter type for baseline

This fixes an issue where we would always evaluate a baseline as a string.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam commented on controllers/application_poller_controller.go:
```diff
@@ -0,0 +1,288 @@
+					p.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to create object", err), activity.URL)
+					return
+				}
+			case err == nil:
```
lets not support updates for this round; 
log a message saying this would have triggered an update, but don't do anything

bradbeam commented on controllers/application_poller_controller.go:
```diff
@@ -0,0 +1,288 @@
+	var exp *optimizev1beta2.Experiment
+	for i := range generatedResources {
+		if expObj, ok := generatedResources[i].(*optimizev1beta2.Experiment); ok {
+			exp = expObj
```
Add annotations to experiment
- experiment url ( from experiments url link + generated name of experiment )
-   - api.RelationExperiments + generatedName

bradbeam commented on controllers/application_poller_controller.go:
```diff
@@ -0,0 +1,288 @@
+	}
+
+	g := &experiment.Generator{
+		Application: app,
```
set experiment name to strings.tolower(ulid.mustnew(xxx))

bradbeam committed:
chore: Fix up additional review comment, upgrade optimize-go version

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

jgustie commented on controllers/application_poller_controller.go:
```diff
@@ -0,0 +1,308 @@
+				exp.ObjectMeta.Annotations = make(map[string]string)
+			}
+
+			exp.ObjectMeta.Annotations[optimizev1beta2.AnnotationExperimentURL] = fmt.Sprintf("%s/%s", experimentURL, exp.ObjectMeta.Name)
```
You can use the `metav1` helper here (it will allocate the missing map for you); also I think there is usually a bit of confusion over trailing slashes in URLs so it might be safer to account for both cases:

```go
metav1.SetMetaDataAnnotation(&exp.ObjectMeta, optimizev1beta2.AnnotationExperimentURL, strings.TrimRight(experimentURL, "/") + "/" + exp.Name)
```

jgustie committed:
Activity errors review

bradbeam committed:
fix: Add missing label for log; update to use metav1 helper.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>

bradbeam committed:
feat: Readd support for updates

This readds support for updating resources we generate. I think the issue I was
initially running in to was because CRDs dont go through the same sanitization process
as core resources, so this means we need to do a little extra legwork.

Signed-off-by: Brad Beam <brad.beam@stormforge.io>
